libraries/aws/common/test/unit/aws_iot_tests_doc_parser.c

```diff
@@ -0,0 +1,271 @@
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file aws_iot_tests_doc_parser.c
+ * @brief Tests for the AWS IoT Parser.
+ */
+
+/* The config header is always included first. */
+#include "iot_config.h"
+
+/* Standard includes. */
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+
+/* AWS IoT parser include. */
+#include "aws_iot_doc_parser.h"
+
+/* Test framework includes. */
+#include "unity_fixture.h"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Wrapper for parsing AWS IoT JSON documents and checking the result.
+ */
+static void _parseJson( bool expectedResult,
+                        const char * pJsonDocument,
+                        size_t jsonDocumentLength,
+                        const char * pJsonKey,
+                        const char * pExpectedJsonValue,
+                        size_t expectedJsonValueLength )
+{
+    const char * pJsonValue = NULL;
+    size_t jsonValueLength = 0;
+
+    TEST_ASSERT_EQUAL_INT( expectedResult,
+                           AwsIotDocParser_FindValue( pJsonDocument,
+                                                      jsonDocumentLength,
+                                                      pJsonKey,
+                                                      strlen( pJsonKey ),
+                                                      &pJsonValue,
+                                                      &jsonValueLength ) );
+
+    if( expectedResult == true )
+    {
+        TEST_ASSERT_EQUAL( expectedJsonValueLength, jsonValueLength );
+        TEST_ASSERT_EQUAL_STRING_LEN( pExpectedJsonValue, pJsonValue, jsonValueLength );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Test group for Shadow parser tests.
+ */
+TEST_GROUP( Aws_Iot_Doc_Unit_Parser );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Test setup for Shadow parser tests.
+ */
+TEST_SETUP( Aws_Iot_Doc_Unit_Parser )
+{
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Test tear down for Shadow parser tests.
+ */
+TEST_TEAR_DOWN( Aws_Iot_Doc_Unit_Parser )
+{
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Test group runner for Shadow parser tests.
+ */
+TEST_GROUP_RUNNER( Aws_Iot_Doc_Unit_Parser )
+{
+    RUN_TEST_CASE( Aws_Iot_Doc_Unit_Parser, JsonValid );
+    RUN_TEST_CASE( Aws_Iot_Doc_Unit_Parser, JsonInvalid );
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Tests parsing valid JSON documents.
+ */
+TEST( Aws_Iot_Doc_Unit_Parser, JsonValid )
+{
+    /* Parse JSON document with string, int, bool, and null. */
+    {
+        const char pJsonDocument[ 82 ] = "{\"name\" \n\r:\n \"John Smith\", \"age\"    :\n\r 30, \n \"isAlive\"  : true, \r \"spouse\":null}";
```

Comment:
The calculation of `jsonDocumentLength` is britttle throughout this file, suggest replacing with `size_t jsonDocumentLength = strlen(pJsonDocument)` in all test cases in which the length matches that of `pJsonDocument`.